### DISCUSSION TITLE
Colorado Inner Radiation Belt Experiment (CIRBE)

### DISCUSSION THREAD

**Discussion by spsvihla** [daniestevez/gr-satellites#423] [Satellite teams]
# Introduction

The CIRBE (Colorado Inner Radiation Belt Experiment) satellite is a 3U (10cm x 10cm x 30cm) CubeSat with the mission of allowing scientists to gain a better understanding of the formation of the inner radiation belt electrons as well as determine where these particles come from and how they behave. The proposed goal of the CIRBE mission is to provide state-of-the-art measurements of .3-3.5 MeV electrons (and 6-35 MeV protons) in the Earth’s radiation belt in a highly inclined Low Earth Orbit.

# Spacecraft Radios

The CIRBE spacecraft contains two radios: one, the UHF radio, for commanding and telemetry, and another, the S-Band radio, for science downlink.

<table>
  <tr>
    <th> Band </th>
    <th> 70 cm (UHF) </th>
    <th> 13 cm (S-Band) </th>
  </tr>
  <tr>
    <td> Mfg./Model </td>
    <td> SpaceQuest UHF TRX-U </td>
    <td> BCT TX Only </td> 
  </tr>
  <tr>
    <td> Frequency </td>
    <td> 437.25 MHz </td>
    <td> 2.402 GHz </td> 
  </tr>
  <tr>
    <td> Modulation </td>
    <td> GFSK </td>
    <td> OQPSK </td> 
  </tr>
  <tr>
    <td> Baudrate </td>
    <td> 9.6k or 19.2k </td>
    <td> 2.0M, 1.0M, or 500k </td> 
  </tr>
  <tr>
    <td> Framing </td>
    <td> AX.25 </td>
    <td> CCSDS </td> 
  </tr>
  <tr>
    <td> Antenna </td>
    <td> Low-gain &lambda;/4 monopole </td>
    <td> High-gain patch </td> 
  </tr>
  <tr>
    <td> Transmit Power </td>
    <td> 2.0 W </td>
    <td> 2.0 W </td> 
  </tr>
</table>

### Framing (AX.25)

All data send by the CIRBE spacecraft over the 70 cm, 437.25 Hz link are contained in AX.25 unnumbered information (UI) frames:

<table>
  <tr>
    <th> </th>
    <th colspan="10"> Field </th>
  </tr>
  <tr>
    <th> </th>
    <th> Flag </th>
    <th> Source </th>
    <th> SSID </th>
    <th> Destination </th>
    <th> SSID </th>
    <th> Control </th>
    <th> PID </th>
    <th> Data </th>
    <th> FCS </th>
    <th> Flag </th> 
  </tr>
  <tr> 
    <td> Size </td>
    <td> 1 </td>
    <td> 6 </td>
    <td> 1 </td>
    <td> 6 </td>
    <td> 1 </td>
    <td> 1 </td>
    <td> 1 </td>
    <td><i> Variable </i></td>_
    <td> 2 </td>
    <td> 1 </td>
  </tr>
  <tr>
    <td> Value </td>
    <td><code> 0x7E </code></td>
    <td><code> 'CIRBE '  </code></td>
    <td><code> 0x60 </code></td>
    <td><code> 'BCT   ' </code></td>
    <td><code> 0x00 </code></td>
    <td><code> 0xFF </code></td>
    <td><code> 0xF0 </code></td>
    <td><i> Variable </i></td>
    <td><i> Variable </i></td>
    <td><code> 0x7E </code></td>
  </tr>
</table>

Note that the control field is set to `0x03` to indicate unnumbered format, and the PID is set to `0xF0` to indicate no layer 3 protocol being used.  

### Source/Destination Callsigns

The AX.25 frames contain the following source and destination callsigns:

<table>
  <tr>
    <th> </th>
    <th colspan="3"> Source </th>
    <th colspan="3"> Destination </th>
  </tr>
  <tr>
    <th> Byte </th>
    <th> Char </th>
    <th> Hex </th>
    <th> Shifted </th>
    <th> Char </th>
    <th> Hex </th>
    <th> Shifted </th>
  </tr>
  <tr>
    <td> 0 </td>
    <td><code> 'C' </code></td>
    <td><code> 0x43 </code></td>
    <td><code> 0x86 </code></td>
    <td><code> 'B' </code></td>
    <td><code> 0x42 </code></td>
    <td><code> 0x84 </code></td>
  </tr>
  <tr>
    <td> 1 </td>
    <td><code> 'I' </code></td>
    <td><code> 0x49 </code></td>
    <td><code> 0x92 </code></td>
    <td><code> 'C' </code></td>
    <td><code> 0x43 </code></td>
    <td><code> 0x86 </code></td>
  </tr>
  <tr>
    <td> 2 </td>
    <td><code> 'R' </code></td>
    <td><code> 0x52 </code></td>
    <td><code> 0xA4 </code></td>
    <td><code> 'T' </code></td>
    <td><code> 0x54 </code></td>
    <td><code> 0xA8 </code></td>
  </tr>
  <tr>
    <td> 3 </td>
    <td><code> 'B' </code></td>
    <td><code> 0x42 </code></td>
    <td><code> 0x84 </code></td>
    <td><code> ' ' </code></td>
    <td><code> 0x20 </code></td>
    <td><code> 0x40 </code></td>
  </tr>
  <tr>
    <td> 4 </td>
    <td><code> 'E' </code></td>
    <td><code> 0x45 </code></td>
    <td><code> 0x8A </code></td>
    <td><code> ' ' </code></td>
    <td><code> 0x20 </code></td>
    <td><code> 0x40 </code></td>
  </tr>
  <tr>
    <td> 5 </td>
    <td><code> ' ' </code></td>
    <td><code> 0x20 </code></td>
    <td><code> 0x40 </code></td>
    <td><code> ' ' </code></td>
    <td><code> 0x20 </code></td>
    <td><code> 0x40 </code></td>
  </tr>
</table>

Note that the 'shifted' column shows the numeric value corresponding to the ASCII character displayed in the 'char' column <i>before</i> it is bit shifted to obtain said character.

### Info Segment

The info segment of every AX.25 frame transmitted by the CIRBE spacecraft begins with a CCSDS header:

<table role="table">
<tbody><tr><th rowspan="2"></th><th rowspan="2">Byte</th><th colspan="8">Description</th></tr>
<tr><th>7</th><th>6</th><th>5</th><th>4</th><th>3</th><th>2</th><th>1</th><th>0</th></tr>
<tr><td rowspan="6"><span>CCSDS Primary Header</span></td><td>0</td><td colspan="3">CCSDS Version (always 0)</td><td>Packet Type (always 0)</td><td>Secondary Header Flag</td><td>Stored Data Flag</td><td colspan="2">APID MSBits</td></tr>
<tr><td>1</td><td colspan="8">APID LSByte</td></tr>
<tr><td>2</td><td colspan="2">Grouping Flag</td><td colspan="6">Sequence Count MSBits</td></tr>
<tr><td>3</td><td colspan="8">Sequence Count LSByte</td></tr>
<tr><td>4</td><td colspan="8">Packet Length MSbyte</td></tr>
<tr><td>5</td><td colspan="8">Packet Length LSbyte</td></tr>
<tr><td rowspan="6"><span>CCSDS Secondary Header</span></td><td>6</td><td colspan="8">Time stamp seconds (J2000 TAI) byte 3 (MSbyte)</td></tr>
<tr><td>7</td><td colspan="8">Time stamp seconds (J2000 TAI) byte 2</td></tr>
<tr><td>8</td><td colspan="8">Time stamp seconds (J2000 TAI) byte 1</td></tr>
<tr><td>9</td><td colspan="8">Time stamp seconds (J2000 TAI) byte 0 (LSbyte)</td></tr>
<tr><td>10</td><td colspan="8">Sub-seconds</td></tr>
<tr><td>11</td><td colspan="8">Reserved (should always equal 0)</td></tr>
</tbody></table>

Interpretation of the following fields is provided:

- <b> Secondary Header Flag </b>: There is no secondary header flag present for packets which span multiple AX.25 frames. For such frames, the secondary header flag is set to `0`.
- <b> Stored Data Flag </b>: This flag is set to `1` if the frame contains stored data, and it is set to `0` if the frame contains real-time data.
- <b> Grouping Flag </b>: When the packet spans multiple frames, this flag is set to `1` for the first frame, it is set to `0` for the middle frames, and it is set to `2` for the final frame. If the packet spans only a single frame, this flag is set to `3`.
- <b> Packet Length </b>: The number of bytes which follow this field, i.e. the info field.
- <b> Sub-seconds </b>: Timestamp in multiples of 200ms, i.e. the flag is set to `0` to indicate 0ms, `1` to indicate 200ms, `2` to indicate 400ms, and so on.

### APIDs

Although the full list is subject to change, the following APIDs are valid:

<table>
  <tr>
    <th> APID </th>
    <th> Meaning </th>
  </tr>
  <tr>
    <td><code> 0x000-0x1FC </code></td>
    <td> XB1 Telemetry Map ID </td>
  </tr>
  <tr>
    <td><code> 0x050 </code></td>
    <td> SOH Packet
  </tr>
  <tr>
    <td><code> 0x1FD </code></td>
    <td> Video Buffer </td>
  </tr>
  <tr>
    <td><code> 0x1FE </code></td>
    <td> Table Data </td>
  </tr>
  <tr>
    <td><code> 0x1FF </code></td>
    <td> Payload Data </td>
  </tr>
</table>

### Beaconing 

The UHF transceiver is configured to beacon at a rate of once every 16 seconds, though this rate is subject to change in the future. 

Each beacon consists of a state of health (SOH) packet, which fits in a single AX.25 frame. These packets have their CCSDS packet length field set to 210 (corresponding to 205 bytes of data following the CCSDS secondary header), their APIDs are set to `0x050`, and their grouping flags are set to `3`.

## Progress So Far

I've created a YAML file and SOH packet parser for CIRBE, and I have pushed my work to [this fork](https://github.com/lasp/gr-satellites) of gr-satellites.

## To-Do

My commit is failing the test "Run daniestevez/gr-satellites-ci-action@v2.0.5". Each of the failures are ModuleNotFoundError and don't seem to have to do with my changes. I'm unsure how to look into the issue because of this.

```
The following tests FAILED:
	  1 - qa_costas_loop_8apsk_cc (Failed)
	  2 - qa_crc (Failed)
	  3 - qa_fixedlen_tagger (Failed)
	  4 - qa_fixedlen_to_pdu (Failed)
	  5 - qa_hdlc (Failed)
	  6 - qa_kiss (Failed)
	  7 - qa_nrzi (Failed)
	  8 - qa_pdu_add_meta (Failed)
	  9 - qa_pdu_head_tail (Failed)
	 10 - qa_pdu_length_filter (Failed)
	 11 - qa_pdu_scrambler (Failed)
	 12 - qa_phase_unwrap (Failed)
	 13 - qa_rs (Failed)
	 14 - qa_viterbi (Failed)
	 15 - qa_ao40_fec_deframer (Failed)
	 16 - qa_usp_deframer (Failed)
	 17 - qa_kiss_server_sink (Failed)
	 18 - qa_telemetry_parser (Failed)
	 19 - qa_rms_agc (Failed)
	 20 - qa_rms_agc_f (Failed)
	 22 - qa_by02 (Failed)
```

Each failure looks similar to this:

```
1: Test command: /usr/bin/sh "qa_costas_loop_8apsk_cc_test.sh"
1: Test timeout computed to be: 10000000
1: Traceback (most recent call last):
1:   File "/github/workspace/python/qa_costas_loop_8apsk_cc.py", line 25, in <module>
1:     from satellites import costas_loop_8apsk_cc
1: ModuleNotFoundError: No module named 'satellites'
 1/22 Test  #1: qa_costas_loop_8apsk_cc ..........***Failed    0.19 sec
test 2
      Start  2: qa_crc
```

**Comment by daniestevez** [comment]
Thank you so much for the detailed information!

Regarding the problem you have with the CI pipeline, the tests give a very unhelpful message due to the way that the Python module is imported, which suppresses/masks the underlying error. Maybe this is something we could improve. A way to get the actual error is to build locally, and from the `build/` directory, run `import python` in a Python interactive prompt. This will give
```
>>> import python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/daniel/gr-satellites/python/__init__.py", line 54, in <module>
    from .adsb_kml import adsb_kml
  File "/home/daniel/gr-satellites/python/adsb_kml.py", line 15, in <module>
    from .telemetry import gomx_3 as tlm
  File "/home/daniel/gr-satellites/python/telemetry/__init__.py", line 28, in <module>
    from .cirbe_70cm import cirbe_70cm
  File "/home/daniel/gr-satellites/python/telemetry/cirbe_70cm.py", line 17, in <module>
    from ax25 import Header
ModuleNotFoundError: No module named 'ax25'
```

The problem is the `from ax25 import Header`. Probably this should read `from .ax25 import Header`.

Other than this, I've taken a quick look at your fork and I think that it will be easy to get the modifications merged upstream. You can submit them in a pull request when they are ready for review.

Do you plan to be working on other features that might be applicable to this mission, such as defragmentation of CCSDS Space Packets or decoding of the OQPSK S-band signal? For defragmentation of Space Packets I'm not sure if we already have a block in gr-satellites, but it would be easy to add one. There isn't an OQPSK demodulator in gr-satellites, but it might be interesting to add one. Lately I've been working with OQPSK because of the Artemis I mission.

**Reply by spsvihla** [reply]
Hi @daniestevez, thanks for the help with that error. I got it fixed and created a [pull request](https://github.com/daniestevez/gr-satellites/pull/426). 

There is interest in building an OQPSK demodulator, but unfortunately our team doesn't have the bandwidth to take on that project now. If we free up and one hasn't been made, we'll be in touch!
